src/cargo/core/compiler/standard_lib.rs

```diff
@@ -0,0 +1,165 @@
+    // lock file from being written, such as setting ephemeral.
+    let std_ws = Workspace::new_virtual(src_path, current_manifest, virtual_manifest, config)?;
+    // `test` is not in the default set because it is optional, but it needs
+    // to be part of the resolve in case we do need it.
+    let mut spec_pkgs = Vec::from(crates);
+    spec_pkgs.push("test".to_string());
+    let spec = Packages::Packages(spec_pkgs);
+    let specs = spec.to_package_id_specs(&std_ws)?;
+    let features = vec!["panic-unwind".to_string(), "backtrace".to_string()];
+    // dev_deps setting shouldn't really matter here.
+    let opts = ResolveOpts::new(
+        /*dev_deps*/ false, &features, /*all_features*/ false,
+        /*uses_default_features*/ true,
+    );
+    let resolve = ops::resolve_ws_with_opts(&std_ws, opts, &specs)?;
+    Ok(resolve)
+}
+
+/// Generate a list of root `Unit`s for the standard library.
+///
+/// The given slice of crate names is the root set.
+pub fn generate_std_roots<'a>(
+    bcx: &BuildContext<'a, '_>,
+    crates: &[String],
+    std_resolve: &'a Resolve,
+) -> CargoResult<Vec<Unit<'a>>> {
+    // Generate the root Units for the standard library.
+    let std_ids = crates
+        .iter()
+        .map(|crate_name| std_resolve.query(crate_name))
+        .collect::<CargoResult<Vec<PackageId>>>()?;
+    // Convert PackageId to Package.
+    let std_pkgs = bcx.packages.get_many(std_ids)?;
+    // Generate a list of Units.
+    std_pkgs
+        .into_iter()
+        .map(|pkg| {
+            let lib = pkg
+                .targets()
+                .iter()
+                .find(|t| t.is_lib())
+                .expect("std has a lib");
+            let unit_for = UnitFor::new_normal();
+            // I don't think we need to bother with Check here, the difference
+            // in time is minimal, and the difference in caching is
+            // significant.
+            let mode = CompileMode::Build;
+            let profile = bcx.profiles.get_profile(
+                pkg.package_id(),
+                /*is_member*/ false,
+                unit_for,
+                mode,
+                bcx.build_config.release,
+            );
+            let features = std_resolve.features_sorted(pkg.package_id());
+            Ok(bcx
+                .units
+                .intern(pkg, lib, profile, Kind::Target, mode, features))
+        })
+        .collect::<CargoResult<Vec<_>>>()
+}
+
+fn detect_sysroot_src_path(ws: &Workspace<'_>) -> CargoResult<PathBuf> {
+    // NOTE: This is temporary until we figure out how to acquire the source.
+    // If we decide to keep the sysroot probe, then BuildConfig will need to
+    // be restructured so that the TargetInfo is created earlier and passed
+    // in, so we don't have this extra call to rustc.
+    let rustc = ws.config().load_global_rustc(Some(ws))?;
+    let output = rustc.process().arg("--print=sysroot").exec_with_output()?;
```

Comment:
We do have a cached input/output procedure with Cargo, and this would actually be pretty nifty to thread through that to ensure that we're not constantly executing `rustc --print sysroot` from within a workspace perhaps?

Reply:
(this would be a good follow-up issue)